numpy/core/function_base.py

```diff
@@ -1,10 +1,24 @@
 from __future__ import division, absolute_import, print_function
 
+import warnings
+import operator
+
 __all__ = ['logspace', 'linspace']
 
 from . import numeric as _nx
 from .numeric import result_type, NaN, shares_memory, MAY_SHARE_BOUNDS, TooHardError
 
+def _index_deprecate(i, stacklevel=2):
```

Comment:
> Note that using index passes numpy booleans, although python booleans always pass as subclasses of int

Good catch :-) gh-7395

> Also, we are counting (integers) not indexing. If we check for indexes we should change the documentation and warning to reflect that requirement.

Some other built-in python functions that are not indexing, but that nonetheless call `__index__`: `hex/oct/bin`, `int(x, base=b)` (calls `x.__int__` but `b.__index__`), `list.__mul__`, `range(...)`, `struct.pack` (when packing an integer), ...

Official documentation of `__index__`: "Called [...] whenever Python needs to losslessly convert the numeric object to an integer object"

The name `__index__` is a historical artifact -- indexing is one place where it's used, but that's not really what it's about. `__index__` is very much the standard Python way to do safe casting to `int` (as opposed to `__int__`, which is for unsafe casting).